src/vos/tests/vts_mvcc.c

```diff
@@ -0,0 +1,562 @@
+}
+
+static int
+punchd_with_flags(struct io_test_args *arg, char *path, daos_epoch_t epoch,
+		  uint64_t flags)
+{
+	struct mvcc_arg	*mvcc_arg = arg->custom;
+	daos_unit_oid_t	 oid;
+	char		 dkey_buf[64];
+	daos_key_t	 dkey = {dkey_buf, sizeof(dkey_buf), 0};
+
+	set_oid(mvcc_arg->i, path, &oid);
+	set_dkey(mvcc_arg->i, path, &dkey);
+
+	flags |= VOS_OF_USE_TIMESTAMPS;
+
+	return vos_obj_punch(arg->ctx.tc_co_hdl, oid, epoch, 0 /* pm_ver */,
+			     flags, &dkey, 0 /* akey_nr */, NULL /* akeys */,
+			     NULL /* dth */);
+}
+
+static int
+punchd_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return punchd_with_flags(arg, path, epoch, 0 /* flags */);
+}
+
+static int
+punchd_dne_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return punchd_with_flags(arg, path, epoch, VOS_OF_COND_PUNCH);
+}
+
+static int
+puncha_with_flags(struct io_test_args *arg, char *path, daos_epoch_t epoch,
+		  uint64_t flags)
+{
+	struct mvcc_arg	*mvcc_arg = arg->custom;
+	daos_unit_oid_t	 oid;
+	char		 dkey_buf[64];
+	daos_key_t	 dkey = {dkey_buf, sizeof(dkey_buf), 0};
+	char		 akey_buf[64];
+	daos_key_t	 akey = {akey_buf, sizeof(akey_buf), 0};
+
+	set_oid(mvcc_arg->i, path, &oid);
+	set_dkey(mvcc_arg->i, path, &dkey);
+	set_akey(mvcc_arg->i, path, &akey);
+
+	flags |= VOS_OF_USE_TIMESTAMPS;
+
+	return vos_obj_punch(arg->ctx.tc_co_hdl, oid, epoch, 0 /* pm_ver */,
+			     flags, &dkey, 1 /* akey_nr */, &akey,
+			     NULL /* dth */);
+}
+
+static int
+puncha_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return puncha_with_flags(arg, path, epoch, 0 /* flags */);
+}
+
+static int
+puncha_ane_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return puncha_with_flags(arg, path, epoch, VOS_OF_COND_PUNCH);
+}
+
+static struct op operations[] = {
+	/* {name,	type,	rlevel,	wlevel,	rtype,	wtype,	func} */
+
+	/* Reads */
+	{"fetch",	T_R,	L_A,	0,	R_R,	0,	NULL},
+	{"listc",	T_R,	L_C,	0,	R_R,	0,	NULL},
+	{"listo",	T_R,	L_O,	0,	R_R,	0,	NULL},
+	{"listd",	T_R,	L_D,	0,	R_R,	0,	NULL},
+	{"queryc",	T_R,	L_C,	0,	R_R,	0,	NULL},
+	{"queryo",	T_R,	L_O,	0,	R_R,	0,	NULL},
+	{"queryd",	T_R,	L_D,	0,	R_R,	0,	NULL},
+
+	/*
+	 * Readwrites
+	 *
+	 *   "_de"	means "if dkey empty"
+	 *   "_dne"	means "if dkey nonempty"
+	 *   "_ae"	means "if akey empty"
+	 *   "_ane"	means "if akey nonempty"
+	 *   "_one"	means "if object nonempty"
+	 *   "_ane"	means "if akey nonempty"
+	 */
+	{"update_de",	T_RW,	L_D,	L_A,	R_E,	W_NE,	update_de_f},
+	{"update_dne",	T_RW,	L_D,	L_A,	R_NE,	W_NE,	update_dne_f},
+	{"update_ae",	T_RW,	L_A,	L_A,	R_E,	W_NE,	update_ae_f},
+	{"update_ane",	T_RW,	L_A,	L_A,	R_NE,	W_NE,	update_ane_f},
+	{"puncho_one",	T_RW,	L_O,	L_O,	R_NE,	W_E,	puncho_one_f},
+	{"punchd_dne",	T_RW,	L_D,	L_D,	R_NE,	W_E,	punchd_dne_f},
+	{"puncha_ane",	T_RW,	L_A,	L_A,	R_NE,	W_E,	puncha_ane_f},
+
+	/* Writes */
+	{"update",	T_W,	0,	L_A,	0,	W_NE,	update_f},
+	{"puncho",	T_W,	0,	L_O,	0,	W_E,	puncho_f},
```

Comment:
Some of the puncho tests are failing with my latest patch.   The checks don't test for emptiness but rather existence.   Empty is harder to test.   It happened to work earlier but not really.  I had a bug that was causing upper trees to get punched.  For instance, punching a dkey caused object to also get punched.   This was introduced in a prior version of my patch when I added code to check existence of parent entities.